projects/compilation.vala

```diff
@@ -0,0 +1,209 @@
+using Gee;
+
+/**
+ * Wraps a Vala.CodeContext and refreshes it as necessary.
+ * TODO: could we implement a GIR documentation symbol resolver by extending this class?
+ */
+class Vls.Compilation : Object {
+    public weak BuildTarget parent_target { get; private set; }
+    private HashSet<string> _packages;
+    // maps URI -> TextDocument
+    private HashMap<string, TextDocument> _sources;
+    private HashSet<string> _vapi_dirs;
+    private HashSet<string> _gir_dirs;
+    private HashSet<string> _metadata_dirs;
+    private HashSet<string> _gresources_dirs;
+    private HashSet<string> _defines;
+
+    // compiler flags
+
+    public bool experimental { get; private set; }
+    public bool experimental_non_null { get; private set; }
+    public Vala.Profile profile { get; private set; }
+    public bool abi_stability { get; private set; }
+
+    // code context
+    /**
+     * Whether the code context needs to be (re-)compiled.
+     */
+    public bool needs_compile { get; private set; default = true; }
+
+    /**
+     * Whether the code context has been updated. This could happen
+     * because we added a new source file, or we changed a source file,
+     * for example.
+     */
+    public bool dirty { get; private set; default = true; }
+
+    private Vala.CodeContext? _ctx;
+
+    public Vala.CodeContext code_context {
+        get {
+            if (dirty) {
+                debug ("dirty context, rebuilding");
+
+                if (_ctx != null) {
+                    // stupid workaround for memory leaks in Vala 0.38
+                    workaround_038 (_ctx, _sources.values);
+                }
+                // generate a new code context 
+                _ctx = new Vala.CodeContext () { keep_going = true };
+                Vala.CodeContext.push (_ctx);
+                dirty = false;
+
+                // string version = Config.libvala_version;
+                _ctx.report = new Reporter ();
+
+                // compiler flags and defines
+                _ctx.profile = profile;
+                switch (profile) {
+                    case Vala.Profile.POSIX:
+                        _ctx.add_define ("POSIX");
+                        break;
+                    default: // case Vala.Profile.GOBJECT:
```

Comment:
Can you explicitly add the `case Vala.Profile.GOBJECT`, and then log a warning (or throw) in `default:` in the case that another profile is added in the future?

Reply:
Logging a message here sounds like a good idea. I want to assume the "right thing" as much as possible.